programs/address-lookup-table/src/processor.rs

```diff
@@ -0,0 +1,377 @@
+
+        let table_account_data_len = LOOKUP_TABLE_META_SIZE;
+        let rent: Rent = invoke_context.get_sysvar(&rent::id())?;
+        let required_lamports = rent
+            .minimum_balance(table_account_data_len)
+            .max(1)
+            .saturating_sub(lookup_table_account.lamports()?);
+
+        if required_lamports > 0 {
+            invoke_context.native_invoke(
+                system_instruction::transfer(&payer_key, &table_key, required_lamports),
+                &[payer_key],
+            )?;
+        }
+
+        invoke_context.native_invoke(
+            system_instruction::allocate(&table_key, table_account_data_len as u64),
+            &[table_key],
+        )?;
+
+        invoke_context.native_invoke(
+            system_instruction::assign(&table_key, &crate::id()),
+            &[table_key],
+        )?;
+
+        let keyed_accounts = invoke_context.get_keyed_accounts()?;
+        let lookup_table_account =
+            keyed_account_at_index(keyed_accounts, first_instruction_account)?;
+        lookup_table_account.set_state(&ProgramState::LookupTable(LookupTableMeta::new(
+            authority_key,
+            derivation_slot,
+        )))?;
+
+        Ok(())
+    }
+
+    fn freeze_lookup_table(
+        invoke_context: &mut InvokeContext,
+        first_instruction_account: usize,
+    ) -> Result<(), InstructionError> {
+        let keyed_accounts = invoke_context.get_keyed_accounts()?;
+
+        let lookup_table_account =
+            keyed_account_at_index(keyed_accounts, first_instruction_account)?;
+        if lookup_table_account.owner()? != crate::id() {
+            return Err(InstructionError::InvalidAccountOwner);
+        }
+
+        let authority_account =
+            keyed_account_at_index(keyed_accounts, checked_add(first_instruction_account, 1)?)?;
+        if authority_account.signer_key().is_none() {
+            return Err(InstructionError::MissingRequiredSignature);
+        }
+
+        let lookup_table_account_ref = lookup_table_account.try_account_ref()?;
+        let lookup_table_data = lookup_table_account_ref.data();
+        let lookup_table = AddressLookupTable::deserialize(lookup_table_data)?;
+
+        if lookup_table.meta.authority.is_none() {
+            ic_msg!(invoke_context, "Lookup table is already frozen");
+            return Err(InstructionError::Immutable);
+        }
+        if lookup_table.meta.authority != Some(*authority_account.unsigned_key()) {
+            return Err(InstructionError::IncorrectAuthority);
+        }
+        if lookup_table.addresses.is_empty() {
+            ic_msg!(invoke_context, "Empty lookup tables cannot be frozen");
+            return Err(InstructionError::InvalidInstructionData);
+        }
+
+        let mut lookup_table_meta = lookup_table.meta;
+        drop(lookup_table_account_ref);
+
+        lookup_table_meta.authority = None;
+        lookup_table_account.set_state(&ProgramState::LookupTable(lookup_table_meta))?;
```

Comment:
nit: this is pedantic, but what do you think about overwriting the rest of the meta portion to `0`s after the authority is set to `None`?  Just in case someone uses a bad deserializer where they only look at those bytes, they could get tricked into thinking there's still an authority.  far-fetched? absolutely. but who knows?

Reply:
I think that's good practice, will do!